### PR TITLE
[W4.3d][F11-A1]Lau Kar Rui

### PR DESCRIPTION
**Pull request by karrui**
Ready for review.

### CODE REVIEW

File: src/seedu/addressbook/data/person/Address.java [hunk 1]
```diff
@@ -8,9 +8,10 @@
  */
 public class Address {
 
-    public static final String EXAMPLE = "123, some street";
-    public static final String MESSAGE_ADDRESS_CONSTRAINTS = "Person addresses can be in any format";
-    public static final String ADDRESS_VALIDATION_REGEX = ".+";
+    public static final String EXAMPLE = "123, some street, some unit number, postal code";
+    private static final String MESSAGE_ADDRESS_CONSTRAINTS = "Address has to be in this format: a/BLOCK, STREET, " +
+            "UNIT, POSTAL_CODE";
+    private static final String ADDRESS_VALIDATION_REGEX = "\\d{1,5}, .+, \\d{1,6}";
```
Comment (PierceAndy) [inline src/seedu/addressbook/data/person/Address.java]: This regex does not work correctly, which would've been caught by tests. What about block numbers with alphabets in them (e.g. 896A)? Are there valid postal codes that aren't 6 digits long? When forming your regex, you can use online regex testers to ensure that your regex works as intended.

File: src/seedu/addressbook/data/person/Address.java [hunk 2]
```diff
@@ -27,7 +30,11 @@ public Address(String address, boolean isPrivate) throws IllegalValueException {
         if (!isValidAddress(trimmedAddress)) {
             throw new IllegalValueException(MESSAGE_ADDRESS_CONSTRAINTS);
         }
-        this.value = trimmedAddress;
+        String[] splitAddress = trimmedAddress.split("\\s*, \\s*");
+        blockNumber = new Block(splitAddress[0]);
+        streetName = new Street(splitAddress[1]);
+        unit = new Unit(splitAddress[2]);
+        postalCode = new PostalCode(splitAddress[3]);
```
Comment (PierceAndy) [inline src/seedu/addressbook/data/person/Address.java]: The regex used by `isValidAddress()` doesn't ensure that the address can be split into 4 elements by the regex in line 33.

File: src/seedu/addressbook/data/person/Address.java [hunk 3]
```diff
@@ -26,34 +35,110 @@ public Address(String address, boolean isPrivate) throws IllegalValueException {
...
     }
 
+    /**
+     * Parses the various Address sub-classes into a String form and returns it.
+     */
     @Override
     public String toString() {
-        return value;
+        return blockNumber.getBlockNumber() + ", " + streetName.getStreetName() + ", " + unit.getUnit() + ", " +
+                postalCode.getPostalCode();
```
Comment (PierceAndy) [inline src/seedu/addressbook/data/person/Address.java]: Consider using StringJoiner: https://docs.oracle.com/javase/8/docs/api/java/util/StringJoiner.html

File: src/seedu/addressbook/data/person/Address.java [hunk 4]
```diff
@@ -8,11 +8,20 @@
  */
 public class Address {
 
-    public static final String EXAMPLE = "123, some street";
-    public static final String MESSAGE_ADDRESS_CONSTRAINTS = "Person addresses can be in any format";
-    public static final String ADDRESS_VALIDATION_REGEX = ".+";
+    public static final String EXAMPLE = "123, some street, some unit number, postal code";
+    private static final String MESSAGE_ADDRESS_CONSTRAINTS = "Address has to be in this format: a/BLOCK, STREET, " +
+            "UNIT, POSTAL_CODE";
+    private static final String ADDRESS_VALIDATION_REGEX = "\\d+,[A-Za-z,\" 0-9]+[A-Za-z,\"0-9],.+, \\d+";
```
Comment (PierceAndy) [inline src/seedu/addressbook/data/person/Address.java]: Would addresses with more than 3 commas pass this validation regex? If they do, would that cause an error down the line?

File: src/seedu/addressbook/data/person/Address.java [hunk 5]
```diff
@@ -26,34 +35,110 @@ public Address(String address, boolean isPrivate) throws IllegalValueException {
...
         return isPrivate;
     }
-}
+
+    /**
+     * Helper class for {@link Address}.
+     * Represents the Block portion of an address.
+     * Note: BlockNumber is stored as a String.
+     */
+    class Block {
```
Comment (PierceAndy) [inline src/seedu/addressbook/data/person/Address.java]: Is there a reason for having no explicit access modifier?

**Comment by PierceAndy** [comment]
Good job on also updating tests, comments, and documents.

@karrui Some comments added. Please close the PR after reading comments.

File: src/seedu/addressbook/data/person/Address.java (hunk 5 above)
Comment (karrui) [reply src/seedu/addressbook/data/person/Address.java]: Good point. Fix included in https://github.com/nus-cs2103-AY1718S1/addressbook-level2/pull/16/commits/9e85c1aa9431f1d290a9bd39066953c3118ee309

File: src/seedu/addressbook/data/person/Address.java (hunk 2 above)
Comment (karrui) [reply src/seedu/addressbook/data/person/Address.java]: More restrictive regex implemented in https://github.com/nus-cs2103-AY1718S1/addressbook-level2/pull/16/commits/932080ea06a8975acc7bfcc1c489f760b08e13e8. Should be fine now.

File: src/seedu/addressbook/data/person/Address.java (hunk 3 above)
Comment (karrui) [reply src/seedu/addressbook/data/person/Address.java]: Seems useful! Implemented with https://github.com/nus-cs2103-AY1718S1/addressbook-level2/pull/16/commits/ca4dab0f7c0ded1c8af90deedf3a463f6a7d8a79 and https://github.com/nus-cs2103-AY1718S1/addressbook-level2/pull/16/commits/4834e1b281a4895b068196b123bce21683ba005f

**Comment by PierceAndy** [comment]
Good job on addressing the problems.

**Pull request closed without merging**